sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -1388,4 +1410,208 @@ public void deidentify(ProcessContext context)
+                                .withOutputTags(FhirIO.Search.OUT, TupleTagList.of(FhirIO.Search.DEAD_LETTER))));
+      }
+
+      /** DoFn for searching messages from the Fhir store with error handling. */
+      static class SearchResourcesFn extends DoFn<KV<String, Map<String, String>>, String> {
+
+        private Counter failedSearches =
+                Metrics.counter(SearchResourcesFn.class, "failed-fhir-searches");
+        private static final Logger LOG = LoggerFactory.getLogger(SearchResourcesFn.class);
+        private final Counter successfulSearches =
+                Metrics.counter(SearchResourcesFn.class, "successful-fhir-searches");
+        private HealthcareApiClient client;
+        private final ValueProvider<String> fhirStore;
+
+        /** Instantiates a new Fhir resources search fn. */
+        SearchResourcesFn(ValueProvider<String> fhirStore) {
+          this.fhirStore = fhirStore;
+        }
+
+        /**
+         * Instantiate healthcare client.
+         *
+         * @throws IOException the io exception
+         */
+        @Setup
+        public void instantiateHealthcareClient() throws IOException {
+          this.client = new HttpHealthcareApiClient();
+        }
+
+        /**
+         * Process element.
+         *
+         * @param context the context
+         */
+        @ProcessElement
+        public void processElement(ProcessContext context) {
+          KV<String, Map<String, String>> elementValues = context.element();
+          try {
+            context.output(searchResources(
+                    this.client, this.fhirStore.toString(), elementValues.getKey(), elementValues.getValue()));
+          } catch (Exception e) {
+            failedSearches.inc();
+            LOG.warn(
+                    String.format(
+                            "Error search FHIR messages writing to Dead Letter "
+                                    + "Queue. Cause: %s Stack Trace: %s",
+                            e.getMessage(), Throwables.getStackTraceAsString(e)));
+            context.output(FhirIO.Search.DEAD_LETTER, HealthcareIOError.of(this.fhirStore.toString(), e));
+          }
+        }
+
+        private String searchResources(HealthcareApiClient client, String fhirStore, String resourceType,
+                                       @Nullable Map<String, String> parameters)
+                throws IllegalArgumentException {
+          long startTime = System.currentTimeMillis();
+
+          HttpHealthcareApiClient.FhirResourcePages.FhirResourcePagesIterator iter =
+                  new HttpHealthcareApiClient.FhirResourcePages.FhirResourcePagesIterator(
+                          client, fhirStore, resourceType, parameters);
+          JsonArray result = new JsonArray();
+          while (iter.hasNext()) {
+            result.addAll(iter.next());
+          }
+          this.successfulSearches.inc();
+          return result.toString();
```

Comment:
I had recommended using async calls. In this case, I see that the calls happen within the iterator. I think this is fine as it is. I have a question though:

Usually, will one element often produce one result from the iterator? Or usually more than one? I see that in lines 1608:1609, for the first iteration, there will be two calls to Fhir (one in hasNext, and one in next). Is it possible to try to reduce this to one?
I expect your connector to have a bottleneck on IO. I would recommend you add a distribution-type metric to track time spent on IO.

Reply:
I think for our use case, usually only one result will be produced from the iterator. Definitely makes sense to avoid one of the two calls, done.

I assume you are talking about adding metrics on our end, but lmk if I misunderstood and there's anything to be added here. Thanks!

Reply:
I think for our use case, usually only one result will be produced from the iterator. Definitely makes sense to avoid one of the two calls, done.

I assume you are talking about adding metrics on our end, but lmk if I misunderstood and there's anything to be added here. Thanks!

Reply:
in this case I mean something like:
```
private Distribution searchLatency =
            Metrics.distribution(SearchResourcesFn.class, "search-latency-or-something-ms");
.....
    long startTime = System.currentTimeMillis();
    ...
    result.addAll(iter.next());
    while (iter.hasNext()) {
      result.addAll(iter.next());
    }
    searchLatency.update(System.currentTimeMillis() - startTime);  // Report latency to Dataflow monitoring for later inspection
```

Reply:
Ah got it, thank you for the pointers! Done.